kmip/services/server/engine.py

```diff
@@ -0,0 +1,404 @@
+from kmip.core.messages.payloads import query
+
+from kmip.core import misc
+
+from kmip.services.server.crypto import engine
+
+
+class KmipEngine(object):
+    """
+    A KMIP request processor that acts as the core of the KmipServer.
+
+    The KmipEngine contains the core application logic for the KmipServer.
+    It processes all KMIP requests and maintains consisten state across
+    client connections.
+
+    Features that are not supported:
+        * KMIP versions > 1.1
+        * Numerous operations, objects, and attributes.
+        * User authentication
+        * Batch processing options: UNDO
+        * Asynchronous operations
+    """
+
+    def __init__(self):
+        """
+        Create a KmipEngine.
+        """
+        self._logger = logging.getLogger(__name__)
+
+        self._cryptography_engine = engine.CryptographyEngine()
+        self._lock = threading.RLock()
+
+        self._id_placeholder = None
+
+        self._protocol_versions = [
+            contents.ProtocolVersion.create(1, 1),
+            contents.ProtocolVersion.create(1, 0)
+        ]
+
+        self._protocol_version = self._protocol_versions[0]
+
+    def _kmip_version_supported(supported):
+        def decorator(function):
+            def wrapper(self, *args, **kwargs):
+                if float(str(self._protocol_version)) < float(supported):
+                    name = function.__name__
+                    operation = ''.join(
+                        [x.capitalize() for x in name[9:].split('_')]
+                    )
+                    raise exceptions.OperationNotSupported(
+                        "{0} is not supported by KMIP {1}".format(
+                            operation,
+                            self._protocol_version
+                        )
+                    )
+                else:
+                    return function(self, *args, **kwargs)
+            return wrapper
+        return decorator
+
+    def _synchronize(function):
+        def decorator(self, *args, **kwargs):
+            with self._lock:
+                return function(self, *args, **kwargs)
+        return decorator
+
+    def _set_protocol_version(self, protocol_version):
+        if protocol_version in self._protocol_versions:
+            self._protocol_version = protocol_version
+        else:
+            raise exceptions.InvalidMessage(
+                "KMIP {0} is not supported by the server.".format(
+                    protocol_version
+                )
+            )
+
+    def _verify_credential(self, request_credential, connection_credential):
+        # TODO (peterhamilton) Add authentication support
+        # 1. If present, verify user ID of connection_credential is valid user.
+        # 2. If present, verify request_credential is valid credential.
+        # 3. If both present, verify that they are compliant with each other.
+        # 4. If neither present, set server to only allow Query operations.
+        pass
+
+    @_synchronize
```

Comment:
`
rlist = [server_socket]
while true:
  sockets_r_ready, , = select.select(rlist, None, None)
  for s in sockets_r_ready:
    if s == server_socket:
      rlist.append(s.connect())
    else:
      msg = read_message(s)
      process_request(msg)
      s.close()
      rlist.remove(s)
  `